src/dict.c

```diff
@@ -675,23 +694,36 @@ void dictRelease(dict *d)
 
 dictEntry *dictFind(dict *d, const void *key)
 {
-    dictEntry *he;
-    uint64_t h, idx, table;
+    dictEntry *he = NULL;
+    uint64_t h, idx;
 
     if (dictSize(d) == 0) return NULL; /* dict is empty */
-    if (dictIsRehashing(d)) _dictRehashStep(d);
+
     h = dictHashKey(d, key);
-    for (table = 0; table <= 1; table++) {
+    idx = h & DICTHT_SIZE_MASK(d->ht_size_exp[0]);
+    if ((long)idx >= d->rehashidx) {
+        he = d->ht_table[0][idx];
+    }
+
+    if (dictIsRehashing(d)) {
+        if ((long)idx >= d->rehashidx && he) {
+            dictBucketRehash(d, h);
+        /* If the 'he' is not in ht0, we perform a random bucket rehash. */
+        } else {
+            assert(!he);
+            _dictRehashStep(d);
+        }
+        /* If rehashing is still ongoing, we lookup in table 1, otherwise in table 0. */
```

Comment:
let's improve the comment.
i.e. the reason is that if we've hit a bucket index that should have been rehashed later, then we re-hashed it right away, so it's in table 1, and if we've hit an index that's already passed rehashing, that's also in 1.
the only case it'll be 0, if if we just completed rehashing and deleted the extra table.
right?

Reply:
Yes, thats right!